File: app/src/main/java/de/dreier/mytargets/features/settings/SettingsActivity.java
/*
 * Copyright (C) 2017 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.features.settings;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.preference.PreferenceScreen;

import de.dreier.mytargets.base.activities.SimpleFragmentActivityBase;
import de.dreier.mytargets.utils.IntentWrapper;
import im.delight.android.languages.Language;

import static android.support.v7.preference.PreferenceFragmentCompat.ARG_PREFERENCE_ROOT;

public class SettingsActivity extends SimpleFragmentActivityBase implements
        PreferenceFragmentCompat.OnPreferenceStartScreenCallback {

    public static IntentWrapper getIntent(ESettingsScreens subScreen) {
        return new IntentWrapper(SettingsActivity.class)
                .with(ARG_PREFERENCE_ROOT, subScreen.getKey());
    }

    @Override
    public Fragment instantiateFragment() {
        String key = getIntent().getStringExtra(ARG_PREFERENCE_ROOT);
        if (key != null) {
            return ESettingsScreens.from(key).create();
        }
        return ESettingsScreens.MAIN.create();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        Language.setFromPreference(this, SettingsManager.KEY_LANGUAGE);
        super.onCreate(savedInstanceState);
        getSupportFragmentManager().addOnBackStackChangedListener(
                () -> {
                    FragmentManager manager = getSupportFragmentManager();
                    if (manager != null) {
                        SettingsFragmentBase currFrag = (SettingsFragmentBase) manager
                                .findFragmentById(android.R.id.content);

                        currFrag.onFragmentResume();
                    }
                });
    }

    @Override
    public boolean onPreferenceStartScreen(PreferenceFragmentCompat preferenceFragmentCompat,
                                           PreferenceScreen preferenceScreen) {
        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        ESettingsScreens screen = ESettingsScreens.from(preferenceScreen.getKey());
        SettingsFragmentBase fragment = screen.create();
        Bundle args = new Bundle();
        args.putString(ARG_PREFERENCE_ROOT, preferenceScreen.getKey());
        fragment.setArguments(args);
        ft.add(android.R.id.content, fragment, preferenceScreen.getKey());
        ft.addToBackStack(preferenceScreen.getKey());
        ft.commit();
        return true;
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        overridePendingTransition(0, 0);
    }
}

File: app/src/main/java/de/dreier/mytargets/features/settings/SettingsActivity.kt
/*
 * Copyright (C) 2018 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.features.settings

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.fragment.app.Fragment
import androidx.fragment.app.transaction
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceFragmentCompat.ARG_PREFERENCE_ROOT
import androidx.preference.PreferenceScreen
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.common.api.ApiException
import de.dreier.mytargets.base.activities.SimpleFragmentActivityBase
import de.dreier.mytargets.features.settings.ESettingsScreens.MAIN
import de.dreier.mytargets.features.settings.backup.provider.GoogleDriveBackup.AsyncRestore.Companion.REQUEST_CODE_SIGN_IN
import timber.log.Timber

class SettingsActivity : SimpleFragmentActivityBase(),
    PreferenceFragmentCompat.OnPreferenceStartScreenCallback {

    public override fun instantiateFragment(): Fragment {
        val key = intent.getStringExtra(ARG_PREFERENCE_ROOT)
        return if (key != null) {
            ESettingsScreens.from(key).create()
        } else MAIN.create()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportFragmentManager.addOnBackStackChangedListener {
            val manager = supportFragmentManager
            if (manager != null) {
                val currFrag = manager
                    .findFragmentById(android.R.id.content) as SettingsFragmentBase
                currFrag.onFragmentResume()
            }
        }
    }

    override fun onPreferenceStartScreen(
        preferenceFragmentCompat: PreferenceFragmentCompat,
        preferenceScreen: PreferenceScreen
    ): Boolean {
        val screen = ESettingsScreens.from(preferenceScreen.key)
        val fragment = screen.create()
        val args = Bundle()
        args.putString(ARG_PREFERENCE_ROOT, preferenceScreen.key)
        fragment.arguments = args
        supportFragmentManager.transaction {
            add(android.R.id.content, fragment, preferenceScreen.key)
            addToBackStack(preferenceScreen.key)
        }
        return true
    }

    override fun onBackPressed() {
        navigationController.finish(animate = false)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                if (childFragment is MainSettingsFragment) {
                    navigationController.finish(animate = false)
                } else {
                    intent.putExtra(ARG_PREFERENCE_ROOT, MAIN.key)
                    val ft = supportFragmentManager.beginTransaction()
                    ft.replace(
                        android.R.id.content,
                        instantiateFragment(),
                        SimpleFragmentActivityBase.FRAGMENT_TAG
                    )
                    ft.commit()
                }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        Timber.d("A.onActivityResult")

        // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
//        if (requestCode == REQUEST_CODE_SIGN_IN) {
//            // The Task returned from this call is always completed, no need to attach
//            // a listener.
//            Timber.d("A.REQUEST_CODE_SIGN_IN")
//            try {
//                val task = GoogleSignIn.getSignedInAccountFromIntent(data)
//                    .getResult(ApiException::class.java)
//                task?.let { handleSignInResult(it) }
//            } catch (e: ApiException) {
//                // The ApiException status code indicates the detailed failure reason.
//                // Please refer to the GoogleSignInStatusCodes class reference for more information.
//                Timber.w("signInResult:failed code=%s", e.statusCode)
//                Timber.w(e)
////            updateUI(null)
//            }
//        }
    }

    private fun handleSignInResult(account: GoogleSignInAccount) {
        Timber.d("signed in")
        // Signed in successfully, show authenticated UI.
//            updateUI(account)
    }
}
